src/GShark/Geometry/Mesh.cs

```diff
@@ -548,6 +548,144 @@ private FaceData CountFaceEdges()
+
+            for (int i = 0; i < 3; i++)
+            {
+                Vector3 w = new Vector3(projection - trianglePoints[i]);
+                Vector3 t = Vector3.CrossProduct(edges_vectors[i], w);
+                double test = n * t;
+                tests.Add(test);
+            }
+
+            if ((tests[0] > 0) && (tests[1] > 0) && (tests[2] > 0))
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        //Function that finds the closer point to a triangle (given the 3 vertices as points)
+        private Point3 ClosestPointToTriangle(List<Point3> trianglePoints, Point3 point)
+        {
+            //Creates Plane from 3 points
+            Plane FacePlane = new Plane(trianglePoints[0], trianglePoints[1], trianglePoints[2]);
+
+            //Find projection of point of the plane
+            Vector3 pointV = new Vector3(point);
+            double cpdistance;
+            Point3 PlaneClosestPoint = FacePlane.ClosestPoint(point, out cpdistance);
+
+            //We create a initial base point "ClosestPoint" that will be overwritten
+            Point3 ClosestPoint = new Point3(0, 0, 0);
+
+            //Checks if the projection of the point on the FacePlane is inside the triangle
+            bool InOrOut = PointInTriangle(PlaneClosestPoint, trianglePoints);
+            if (InOrOut == true)
+            {
+                //Checks if the projection of the point on the FacePlane is inside the triangle
+                ClosestPoint = PlaneClosestPoint;
+            }
+
+            else //If not, we need to perform a few more calculations.
+                 //We create a line(a Rhino line, a mathematical segment)
+                 //for each of the edges and then find the closest point on each of the edges,
+                 //and then take the closest one out of these 3
+            {
+                //Create a list to store the closest point to each of the edges
+                List<Point3> EdgesClosestPoints = new List<Point3>();
+                Line segment_AB = new Line(trianglePoints[0], trianglePoints[1]);
+                Line segment_BC = new Line(trianglePoints[1], trianglePoints[2]);
+                Line segment_CA = new Line(trianglePoints[2], trianglePoints[0]);
+                List<Line> Segments_list = new List<Line>() { segment_AB, segment_BC, segment_CA };
+
+                //Create a segment for each vertex
+                foreach (Line segment in Segments_list)
+                {
+                    Point3 SegmentClosestPoint = segment.ClosestPoint(point);
+                    EdgesClosestPoints.Add(SegmentClosestPoint);
+                }
+                Point3 edgeClosestPoint = CloudClosestPoint(point, EdgesClosestPoints);
+                ClosestPoint = edgeClosestPoint;
+            }
+
+            return ClosestPoint;
+        }
+
+        public Point3 MeshClosestPointTest(Point3 point)
```

Comment:
I think this can start to be named just `ClosestPoint` right?

So we will call `Point3 closestPoint = myMesh.ClosestPoint(testpoint)`

Reply:
Each closest point to a Face I called ClosestPoint that's why I needed another name to call the closest out of all.

Reply:
mmh. I'm not convinced yet :)
I'm talking about the method name, line 647.

Reply:
Oh, I see... you mean the name of the method. Yes, that can be shorter for sure.